### PR TITLE
Use Optionals To Handle and Filter Out Ghosts

### PR DESCRIPTION
**Pull request by fppt**

### CODE REVIEW

File: grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipEdge.java [hunk 1]
```diff
@@ -62,8 +63,15 @@
     private final Cache<Role> valueRole = new Cache<>(Cacheable.concept(), () -> edge().tx().getSchemaConcept(LabelId.of(
             edge().property(Schema.EdgeProperty.RELATIONSHIP_ROLE_VALUE_LABEL_ID))));
 
-    private final Cache<Thing> owner = new Cache<>(Cacheable.concept(), () -> edge().tx().factory().buildConcept(edge().source()));
-    private final Cache<Thing> value = new Cache<>(Cacheable.concept(), () -> edge().tx().factory().buildConcept(edge().target()));
+    private final Cache<Thing> owner = new Cache<>(Cacheable.concept(), () -> {
+        VertexElement vertexElement = edge().source().orElseThrow(() -> GraknTxOperationException.missingOwner(getId()));
+        return edge().tx().factory().<Thing>buildConcept(vertexElement).orElseThrow(() -> GraknTxOperationException.missingOwner(getId()));
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/concept/RelationshipEdge.java]: Desperate attempt to improve the caches everywhere:
```
return edge().source().flatMap(edge().tx().factory()::buildConcept).orElseThrow(() ->
    GraknTxOperationException.missingOwner(getId())
);
```

(feel free to extract more variables to make it clearer or w/e)

File: grakn-kb/src/main/java/ai/grakn/kb/internal/concept/TypeImpl.java [hunk 2]
```diff
@@ -241,7 +248,9 @@ boolean deletionAllowed(){
 
     Stream<V> instancesDirect(){
         return vertex().getEdgesOfType(Direction.IN, Schema.EdgeLabel.SHARD).
-                map(edge -> vertex().tx().factory().buildShard(edge.source())).
+                map(EdgeElement::source).
+                flatMap(CommonUtil::optionalToStream).
+                map(source -> vertex().tx().factory().buildShard(source)).
```
Comment (aelred) [inline grakn-kb/src/main/java/ai/grakn/kb/internal/concept/TypeImpl.java]: I _think_ these kinds of lambdas can be made into method references like this:
```
map(vertex().tx().factory()::buildShard)
```
(although remember `vertex().tx().factory()` would now be called outside the lambda)

Comment (fppt) [reply grakn-kb/src/main/java/ai/grakn/kb/internal/concept/TypeImpl.java]: It's can't in this case because you passing in `edge.source()` not `edge`

**Comment by fppt** [comment]
!rtg

**Comment by fppt** [comment]
!rtg - ai.grakn.test.engine.postprocessing.PostProcessingIT.checkThatDuplicateResourcesAtLargerScaleAreMerged

**Pull request merged**